File: fineTuneSLM.py
import argparse
import json
import os
import logging
from sklearn.model_selection import train_test_split
from transformers import DistilBertTokenizerFast, DistilBertForQuestionAnswering, Trainer, TrainingArguments
import torch
from torch.utils.data import Dataset
from tqdm import tqdm

logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s")
logger = logging.getLogger(__name__)


class MathQADataset(Dataset):
    def __init__(self, qa_pairs, tokenizer, max_length=512):
        self.tokenizer = tokenizer
        self.max_length = max_length
        self.encodings = []
        self.start_positions = []
        self.end_positions = []
        self.raw_pairs = []

        for qa_pair in tqdm(qa_pairs, desc="Tokenizing"):
            question = qa_pair["question"]
            context = qa_pair["context"]
            answer = qa_pair["answer"]

            # Force offset mapping with fast tokenizer
            encoding = tokenizer(
                text=question,
                text_pair=context,
                max_length=max_length,
                padding="max_length",
                truncation=True,
                return_tensors="pt",
                return_offsets_mapping=True,
            )

            start_pos, end_pos = self._find_answer_positions(context, answer, encoding)

            if start_pos == 0 and end_pos == 0 and answer.strip():
                continue

            self.encodings.append({
                "input_ids": encoding["input_ids"].squeeze(0),
                "attention_mask": encoding["attention_mask"].squeeze(0),
            })
            self.start_positions.append(start_pos)
            self.end_positions.append(end_pos)
            self.raw_pairs.append(qa_pair)

    def __len__(self):
        return len(self.encodings)

    def __getitem__(self, idx):
        item = {k: v for k, v in self.encodings[idx].items()}
        item["start_positions"] = self.start_positions[idx]
        item["end_positions"] = self.end_positions[idx]
        return item

    @staticmethod
    def _find_answer_positions(context, answer, encoding):
        offset_mapping = encoding["offset_mapping"][0].tolist()
        start_char = context.find(answer)
        if start_char == -1:
            return 0, 0
        end_char = start_char + len(answer) - 1

        start_token = end_token = None
        for idx, (ts, te) in enumerate(offset_mapping):
            if ts == te:
                continue
            if start_token is None and ts <= start_char < te:
                start_token = idx
            if ts <= end_char < te:
                end_token = idx
                break
        return start_token or 0, end_token or 0


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("-i", "--input-dir", required=True)
    parser.add_argument("-t", "--tmp-dir", required=True)
    parser.add_argument("-w", "--output-dir", required=True)
    parser.add_argument("--num-epochs", type=int, default=1)
    args = parser.parse_args()

    os.makedirs(args.tmp_dir, exist_ok=True)
    os.makedirs(args.output_dir, exist_ok=True)

    device = torch.device("cpu")

    # Load sample data
    sample_path = os.path.join(args.input_dir, "sample_qa.json")
    with open(sample_path, "r", encoding="utf-8") as f:
        data = json.load(f)
    qa_pairs = data.get("qa_pairs", [])

    model_name = "distilbert-base-uncased"
    tokenizer = DistilBertTokenizerFast.from_pretrained(model_name)
    model = DistilBertForQuestionAnswering.from_pretrained(model_name).to(device)

    train_pairs, val_pairs = train_test_split(qa_pairs, test_size=0.2, random_state=42)

    train_dataset = MathQADataset(train_pairs, tokenizer)
    val_dataset = MathQADataset(val_pairs, tokenizer)

    training_args = TrainingArguments(
        output_dir=args.tmp_dir,
        num_train_epochs=args.num_epochs,
        per_device_train_batch_size=4,
        per_device_eval_batch_size=4,
        logging_steps=10,
        save_strategy="no",
        eval_strategy="no",
        report_to="none",
    )

    trainer = Trainer(
        model=model,
        args=training_args,
        train_dataset=train_dataset,
        eval_dataset=val_dataset,
    )

    logger.info("Starting fine-tuning on %d samples...", len(train_dataset))
    trainer.train()

    model.save_pretrained(args.output_dir)
    tokenizer.save_pretrained(args.output_dir)
    logger.info("Training completed and model saved!")


if __name__ == "__main__":
    main()

File: fineTuneSLM.cpp
#include <torch/torch.h>
#include <nlohmann/json.hpp>
#include <filesystem>
#include <fstream>
#include <vector>
#include <string>
#include <iostream>
#include <stdexcept>
#include <tqdm/tqdm.h> // Assumes a C++ tqdm equivalent or custom progress bar

using json = nlohmann::json;
namespace fs = std::filesystem;

// Simple struct to hold QA pair
struct QAPair {
    std::string question;
    std::string answer;
    std::string context;
};

// Simple tokenizer (placeholder for DistilBERT tokenizer)
class SimpleTokenizer {
public:
    SimpleTokenizer(const std::string& vocab_path) {
        // Load vocabulary (simplified; assumes vocab.txt in -w directory)
        std::ifstream vocab_file(vocab_path);
        std::string word;
        int id = 0;
        while (vocab_file >> word) {
            vocab[word] = id++;
        }
    }

    std::vector<int64_t> tokenize(const std::string& text) {
        std::vector<int64_t> tokens;
        // Simplified: split by space and map to vocab IDs
        std::istringstream iss(text);
        std::string word;
        while (iss >> word) {
            auto it = vocab.find(word);
            tokens.push_back(it != vocab.end() ? it->second : 0); // 0 for unknown
        }
        return tokens;
    }

private:
    std::map<std::string, int64_t> vocab;
};

// Custom dataset for QA
class MathQADataset : public torch::data::Dataset<MathQADataset> {
public:
    MathQADataset(const std::vector<QAPair>& qa_pairs, SimpleTokenizer& tokenizer, size_t max_length = 512)
        : qa_pairs_(qa_pairs), tokenizer_(tokenizer), max_length_(max_length) {}

    torch::data::Example<> get(size_t index) override {
        const auto& qa = qa_pairs_[index];
        auto question_tokens = tokenizer_.tokenize(qa.question + " [SEP] " + qa.context);
        auto answer_tokens = tokenizer_.tokenize(qa.answer);

        // Pad or truncate to max_length
        std::vector<int64_t> input_ids(max_length_, 0); // 0 for padding
        std::vector<int64_t> attention_mask(max_length_, 0);
        size_t len = std::min(question_tokens.size(), max_length_);
        for (size_t i = 0; i < len; ++i) {
            input_ids[i] = question_tokens[i];
            attention_mask[i] = 1;
        }

        // Find answer positions (simplified)
        int64_t start_pos = 0, end_pos = 0;
        for (size_t i = 0; i <= input_ids.size() - answer_tokens.size(); ++i) {
            bool match = true;
            for (size_t j = 0; j < answer_tokens.size(); ++j) {
                if (i + j >= input_ids.size() || input_ids[i + j] != answer_tokens[j]) {
                    match = false;
                    break;
                }
            }
            if (match) {
                start_pos = i;
                end_pos = i + answer_tokens.size() - 1;
                break;
            }
        }

        return {
            torch::tensor(input_ids),
            {
                torch::tensor(attention_mask),
                torch::tensor(start_pos),
                torch::tensor(end_pos)
            }
        };
    }

    torch::optional<size_t> size() const override {
        return qa_pairs_.size();
    }

private:
    std::vector<QAPair> qa_pairs_;
    SimpleTokenizer& tokenizer_;
    size_t max_length_;
};

// Simple transformer model (placeholder for DistilBERT)
struct SimpleTransformer : torch::nn::Module {
    SimpleTransformer() {
        // Placeholder: Define a simple transformer architecture
        linear = register_module("linear", torch::nn::Linear(512, 2)); // Outputs start/end logits
    }

    torch::Tensor forward(torch::Tensor input) {
        // Placeholder: Simplified forward pass
        return linear->forward(input);
    }

    torch::nn::Linear linear{nullptr};
};

std::vector<QAPair> load_qa_pairs(const std::string& input_dir) {
    std::vector<QAPair> qa_pairs;
    for (const auto& entry : fs::directory_iterator(input_dir)) {
        if (entry.path().extension() == ".json") {
            std::ifstream file(entry.path());
            json data;
            file >> data;
            for (const auto& qa : data["qa_pairs"]) {
                qa_pairs.push_back({
                    qa["question"].get<std::string>(),
                    qa["answer"].get<std::string>(),
                    qa["context"].get<std::string>()
                });
            }
        }
    }
    return qa_pairs;
}

int main(int argc, char* argv[]) {
    // Parse arguments
    std::string input_dir, tmp_dir, output_dir;
    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        if (arg == "-I" && i + 1 < argc) input_dir = argv[++i];
        else if (arg == "-t" && i + 1 < argc) tmp_dir = argv[++i];
        else if (arg == "-w" && i + 1 < argc) output_dir = argv[++i];
        else {
            std::cerr << "Usage: " << argv[0] << " -I <input_dir> -t <tmp_dir> -w <output_dir>" << std::endl;
            return 1;
        }
    }

    if (input_dir.empty() || tmp_dir.empty() || output_dir.empty()) {
        std::cerr << "All arguments (-I, -t, -w) are required" << std::endl;
        return 1;
    }

    // Create directories
    fs::create_directories(output_dir);
    fs::create_directories(tmp_dir);

    // Check MPS availability
    torch::Device device(torch::kMPS);
    if (!torch::hasMPS()) {
        std::cout << "MPS not available, falling back to CPU" << std::endl;
        device = torch::kCPU;
    } else {
        std::cout << "Using device: MPS" << std::endl;
    }

    // Load tokenizer (assumes vocab.txt in output_dir)
    SimpleTokenizer tokenizer(output_dir + "/vocab.txt");

    // Load QA pairs
    std::cout << "Loading QA pairs..." << std::endl;
    auto qa_pairs = load_qa_pairs(input_dir);
    std::cout << "Loaded " << qa_pairs.size() << " QA pairs" << std::endl;

    // Create dataset and dataloader
    auto dataset = MathQADataset(qa_pairs, tokenizer).map(torch::data::transforms::Stack<>());
    auto dataloader = torch::data::make_data_loader(
        std::move(dataset),
        torch::data::DataLoaderOptions().batch_size(8).workers(0) // No workers for macOS
    );

    // Load or download model (assumes distilbert.pt exists or is downloaded)
    std::string model_path = output_dir + "/distilbert.pt";
    SimpleTransformer model;
    if (fs::exists(model_path)) {
        std::cout << "Loading pre-trained model from " << model_path << std::endl;
        torch::load(model, model_path);
    } else {
        std::cout << "Pre-trained model not found at " << model_path << ". Initialize new model." << std::endl;
        // In practice, download or convert from Python model
    }
    model.to(device);

    // Optimizer
    torch::optim::Adam optimizer(model.parameters(), torch::optim::AdamOptions(1e-5));

    // Training loop
    std::cout << "Starting fine-tuning..." << std::endl;
    model.train();
    for (int epoch = 0; epoch < 3; ++epoch) {
        for (auto& batch : *dataloader) {
            auto input_ids = batch.data.to(device);
            auto attention_mask = batch.target[0].to(device);
            auto start_positions = batch.target[1].to(device);
            auto end_positions = batch.target[2].to(device);

            optimizer.zero_grad();
            auto outputs = model(input_ids);
            auto loss = torch::cross_entropy(outputs.slice(1, 0, 1).squeeze(1), start_positions) +
                        torch::cross_entropy(outputs.slice(1, 1, 2).squeeze(1), end_positions);
            loss.backward();
            optimizer.step();

            std::cout << "Epoch " << epoch + 1 << ", Loss: " << loss.item<float>() << std::endl;
        }

        // Save checkpoint
        std::string checkpoint_path = tmp_dir + "/checkpoint_epoch_" + std::to_string(epoch + 1) + ".pt";
        torch::save(model, checkpoint_path);
        std::cout << "Saved checkpoint to " << checkpoint_path << std::endl;
    }

    // Save final model
    std::cout << "Saving fine-tuned model to " << model_path << std::endl;
    torch::save(model, model_path);

    return 0;
}